### PR TITLE
feat(integrations): add Semrush

### PR DESCRIPTION
**Pull request by waleedlatif1**
## Summary
- Adds Semrush as an API-key block with 44 operations covering the SEO API end to end: overview (domain/subdomain/URL totals, history, all-database rollups), domain and subdomain organic/paid keywords, competitors, ad copies, ad history and PLA reports, URL reports, keyword research (overview, batch, difficulty, related, broad match, questions, organic/paid SERP results, ad history), Domain vs. Domain comparison, and backlinks (overview, backlinks, referring domains and IPs, TLD/country distribution, anchors, indexed pages, competitors)
- Reports come back as delimited CSV, so the shared decoder maps each header cell back to the export column it was requested as instead of reading by position — the API can return fewer columns than were asked for (`phrase_these` accepts `Td` and omits Trends), and `Fk`/`Fp` both render under the header label `SERP Features`
- A rejected request arrives as HTTP 200 with an `ERROR nnn :: ...` body, so the body is checked even on success; `export_escape=1` is always set and the parser is quote-aware so titles, anchors, and ad copy containing the delimiter survive
- Row limits are clamped so a single step cannot pull an unbounded result into memory
- Every column, parameter, and response field traces to a verbatim example in the Semrush API reference — nothing inferred

## Not included
- Trends, Projects (Site Audit, Position Tracking), and Listing Management are separate Semrush products with their own endpoints and subscriptions
- `backlinks_matrix` and `backlinks_comparison` take repeated array parameters that do not fit the shared request builder
- `subdomain_organic_unique` documents a response format inconsistent with every other report

## Type of Change
- [x] New feature (non-breaking change which adds functionality)

## Testing
- 27 unit tests over the decoder, request building, and block/tool alignment, using the API reference examples verbatim as fixtures. Reverting the header-mapping fix turns 3 of them red, including one that decodes a report with a middle column removed
- One test walks all 44 registered tools and asserts every export column they request resolves; another asserts, per operation, that the block shows a field for exactly the params its tool declares and marks the required ones required
- \`bun run type-check\`, \`bun run lint\`, and \`bun run check:audits\` (33/33) pass

## Checklist
- [x] Code follows project style guidelines
- [x] Self-reviewed my changes
- [x] Tests added/updated and passing
- [x] No new warnings introduced
- [x] I confirm that I have read and agree to the terms outlined in the [Contributor License Agreement (CLA)](./CONTRIBUTING.md#contributor-license-agreement-cla)

### CODE REVIEW

File: apps/sim/tools/semrush/semrush.test.ts [hunk 1]
```diff
@@ -0,0 +1,450 @@
...
+  })
+
+  it('surfaces a non-2xx body', async () => {
+    await expect(
+      semrushDomainOverviewTool.transformResponse!(csvResponse('ERROR 50 :: NOTHING FOUND', 400))
+    ).rejects.toThrow('ERROR 50 :: NOTHING FOUND')
+  })
+
+  it('requests the columns it decodes', () => {
+    const url = semrushDomainOverviewTool.request.url as (params: any) => string
```
[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/sim/tools/semrush/semrush.test.ts]: Fixed. Replaced the four erased casts with a typed `requestUrl<P>(tool, params)` helper that infers each tool's own parameter type and narrows the `string | ((params: P) => string)` union, so a request-contract change now breaks the fixture at compile time instead of silently passing.

File: apps/sim/tools/semrush/domain_vs_domain.ts [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+      })
+    },
+    method: 'GET',
+    headers: () => ({ Accept: 'text/csv' }),
+  },
+
+  transformResponse: async (response: Response, params) => {
+    const { headers, rows } = await readSemrushReportWithHeader(response)
+    const domainCount = parseDomains(params?.domains ?? '').length
+    const comparedDomains = headers.slice(1, 1 + domainCount)
```
[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/sim/tools/semrush/domain_vs_domain.ts]: Valid, and it was the one place left decoding by position — fixed. The metric columns are now located by their own headers (`Competition`, `Search Volume`, `CPC`), and the compared domains are whatever sits between the keyword column and the first metric, so a dropped position column shortens that run instead of pushing a metric into it. An omitted metric column resolves to null rather than reading its neighbour. Two tests cover both cases and fail against the old slice.

File: apps/sim/tools/semrush/domain_organic_competitors.ts [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+          },
+          organicCost: {
+            type: 'number',
+            description: 'Estimated Google Ads price of the organic keywords',
+            optional: true,
+            nullable: true,
+          },
+          paidKeywords: {
+            type: 'number',
+            description: 'Keywords the target is buying in Google Ads',
```
[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/sim/tools/semrush/domain_organic_competitors.ts]: Correct — in `domain_organic_organic` every row is a competing domain, so `Or`/`Ot`/`Oc`/`Ad` are that competitor's metrics, not the target's. Fixed here and in the paid-competitor report, and the same class of error in Winners and Losers and Top Domains where each row is its own domain.

File: apps/sim/tools/semrush/subdomain_paid_keywords.ts [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+          },
+          trafficPercent: {
+            type: 'number',
+            description: 'Share of the traffic to the target driven by this keyword',
+            optional: true,
+            nullable: true,
+          },
+          trafficCost: {
+            type: 'number',
+            description: 'Share of the traffic cost of the target attributed to this keyword',
```
[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/sim/tools/semrush/subdomain_paid_keywords.ts]: Correct, and the docs confirm it: the shared column table defines `Tc` as a percentage, but the response header differs per report — `domain_organic` returns `Traffic Cost (%)` with values like 44.40, while `domain_adwords`, `url_organic`, and `url_adwords` return `Traffic Cost` with absolute values (5976, 78114). Reworded to an estimated cost on the paid and URL reports and left as a percentage share on the organic domain/subdomain reports, matching each one's documented example.

File: apps/sim/tools/semrush/keyword_overview.ts [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+  id: 'semrush_keyword_overview',
+  name: 'Semrush Keyword Overview',
+  description: 'Get search volume, CPC, competition, difficulty, and intent for a single keyword.',
+  version: '1.0.0',
+
+  params: {
+    apiKey: {
+      type: 'string',
+      required: true,
+      visibility: 'user-only',
```
[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/sim/tools/semrush/keyword_overview.ts:24]: Agreed, and the codebase backs it: across `apps/sim/tools`, `user-or-llm` outnumbers `user-only` roughly 10:1 for limit (466/43), offset (124/36), sort (327/11), and date-like params (182/12), and the add-integration guidance reserves `user-only` for API keys, credentials, and account-specific IDs. Every Semrush tool now has exactly one `user-only` param — `apiKey` — with target scope, limit, offset, date, daily, sort, and filter all `user-or-llm`.

File: apps/sim/tools/semrush/backlinks_anchors.ts [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+  name: 'Semrush Backlink Anchors',
+  description:
+    'List the anchor texts used in backlinks to a target, with domain and backlink counts.',
+  version: '1.0.0',
+
+  params: {
+    apiKey: {
+      type: 'string',
+      required: true,
+      visibility: 'user-only',
```
[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/sim/tools/semrush/backlinks_anchors.ts:30]: Agreed, and the codebase backs it: across `apps/sim/tools`, `user-or-llm` outnumbers `user-only` roughly 10:1 for limit (466/43), offset (124/36), sort (327/11), and date-like params (182/12), and the add-integration guidance reserves `user-only` for API keys, credentials, and account-specific IDs. Every Semrush tool now has exactly one `user-only` param — `apiKey` — with target scope, limit, offset, date, daily, sort, and filter all `user-or-llm`.

File: apps/sim/tools/semrush/backlinks_geo_distribution.ts [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+  id: 'semrush_backlinks_geo_distribution',
+  name: 'Semrush Backlink Country Distribution',
+  description: 'Break a target backlink profile down by the country of its referring domains.',
+  version: '1.0.0',
+
+  params: {
+    apiKey: {
+      type: 'string',
+      required: true,
+      visibility: 'user-only',
```
[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/sim/tools/semrush/backlinks_geo_distribution.ts:29]: Agreed, and the codebase backs it: across `apps/sim/tools`, `user-or-llm` outnumbers `user-only` roughly 10:1 for limit (466/43), offset (124/36), sort (327/11), and date-like params (182/12), and the add-integration guidance reserves `user-only` for API keys, credentials, and account-specific IDs. Every Semrush tool now has exactly one `user-only` param — `apiKey` — with target scope, limit, offset, date, daily, sort, and filter all `user-or-llm`.

File: apps/sim/tools/semrush/utils.ts [hunk 8]
```diff
@@ -0,0 +1,334 @@
...
+    url.searchParams.set(name, String(value))
+  }
+  return url.toString()
+}
+
+/** Clamps a user- or model-supplied row limit to the range a workflow can hold. */
+export function normalizeLimit(limit: number | string | undefined, fallback: number): number {
+  const parsed = Number(limit)
+  if (!Number.isFinite(parsed) || parsed <= 0) return fallback
+  return Math.min(Math.floor(parsed), SEMRUSH_MAX_ROWS)
```
[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/sim/tools/semrush/utils.ts]: Real edge case — fixed. `0.5` passed the `<= 0` guard and floored to `0`, sending `display_limit=0`, which is a different request from the one the caller made. The floor is now held at one row, with a test that fails against the old expression.

File: apps/sim/tools/semrush/domain_overview_history.ts [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+    return {
+      success: true,
+      output: { history: rows },
+    }
+  },
+
+  outputs: {
+    history: {
+      type: 'array',
+      description: 'Historical data points, newest first',
```
[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/sim/tools/semrush/domain_overview_history.ts]: Correct — the tool exposes `display_sort`, so `dt_asc` makes the claim false. Dropped the ordering claim from all three history reports (domain, subdomain, URL); they now describe the rows as being in the requested sort order.

File: apps/sim/tools/semrush/domain_paid_competitors.ts [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+          },
+          commonKeywords: {
+            type: 'number',
+            description: 'Keywords both domains rank for in the top 100 organic results',
+            optional: true,
+            nullable: true,
+          },
+          paidKeywords: {
+            type: 'number',
+            description: 'Keywords the target is buying in Google Ads',
```
[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/sim/tools/semrush/domain_paid_competitors.ts]: Same finding as the organic-competitor report and fixed with it: in `domain_adwords_adwords` each row is a competing domain, so `Ad` is that competitor's paid-keyword count. All the per-row metrics in both competitor reports now name the competing domain.

**Comment by waleedlatif1** [comment]
@greptile
